information/specification.md

```diff
@@ -0,0 +1,706 @@
+
+### Transaction Output
+
+A _transaction output_ is a pair of values (**_t_**, **_a_**), where:
+
+  * **_t_** is a [target address](#address);
+  * **_a_** is a [coin amount](#coin-amount).
+
+# Interface
+
+All coin selection algorithms used by Cardano Wallet implement a _common
+interface_.
+
+Essentially, a coin selection algorithm is simply a _mathematical function_
+that, when applied to a [standard set of parameters](#parameters), will produce
+a [standard set of results](#results).
+
+This section defines the [parameters](#parameters) accepted by coin selection
+algorithms used in Cardano Wallet, the [results](#results) returned, as well as
+the various [error conditions](#failure-modes) that may occur.
+
+In this section, the terms _coin selection algorithm_ and _coin selection
+function_ will be used interchangeably.
+
+## Parameters
+
+A coin selection function accepts the following parameters:
+
+ 1. #### Requested Output List
+
+    A list of payments to be made to recipient addresses, encoded as a list of
+    [transaction outputs](#transaction-output).
+
+ 2. #### Initial UTxO Set
+
+    A subset of a wallet's [UTxO set](#utxo-set).
+
+    The coin selection algorithm will select entries from within this set in
+    order to cover payments listed in the [requested output
+    list](#requested-output-list).
+
+    Normally, this parameter would be assigned with the complete [UTxO
+    set](#utxo-set) of a wallet, giving the coin selection algorithm access to
+    the total value associated with that wallet.
+
+ 3. #### Maximum Input Count
+
+    An _upper bound_ on the number of UTxO entries that the coin selection
+    algorithm is permitted to select from the [initial UTxO
+    set](#initial-utxo-set).
+
+## Results
+
+A coin selection function produces the following result values:
+
+ 1. #### Coin Selection
+
+    The _coin selection_ is a record with three fields:
+
+      * A set of **_inputs_**, equivalent to a subset of the
+        [initial UTxO set](#initial-utxo-set).
+
+        Represents the value that has been selected from the wallet in order to
+        cover the total payment value.
+
+      * A set of **_outputs_** (see [transaction output](#transaction-output)).
+
+        Represents the set of payments to be made to recipient addresses.
+
+      * A set of **_change values_**, where each change value is simply a
+        [coin amount](#coin-amount).
+
+        Represents the change to be returned to the wallet.
+
+ 2. #### Remaining UTxO Set
+
+    The _remaining UTxO set_ is a subset of the [initial UTxO
+    set](#initial-utxo-set).
+
+    It represents the set of values that remain after the coin selection
+    algorithm has removed values to pay for entries in the [requested output
+    list](#requested-output-list).
+
+    If a coin selection algorithm is applied to the _complete_ UTxO set of a
+    wallet, then the remaining UTxO set represents the _updated_ UTxO set of the
+    wallet.
+
+## Properties
+
```

Comment:
I wonder if this section needs revision so that it's more human readable.

Will we also have a formal specification of coin selection?

Reply:
@rvl wrote:
> I wonder if this section needs revision so that it's more human readable.

Thanks, this kind of feedback is very useful.

I'm currently attempting to write each property definition so that it has _both_ a human-readable explanation (with natural language), _and_ a more mathematical definition underneath. This way, people who are not interested in mathematics will still be able to gain an intuition of each property, but those who require more detail can read the mathematical definitions.

Do you think the human-readable explanations are accessible enough?

My interpretation of the [requirements for this specification](https://jira.iohk.io/browse/ADP-189) is that someone who's never seen the Haskell code _should_ be able to read it and then be able to re-implement these algorithms (in their language of choice). Of course, with a human language explanation, it's possible that we'll miss something. But I think it's okay for us to make further revisions to the specification if people identify things that are not clear enough.

> Will we also have a formal specification of coin selection?

I'm not sure of the answer to that! @KtorZ, are you able to shed some light on this area?

Reply:
Closing this section after internal discussion with @rvl, as it seems we really do need to have both of the following:

* **human-language explanations**, to help casual readers gain an intuition of what each property is for.
* **mathematical laws**, to help people that want to re-implement these algorithms.

It may be the case that in future we'll write a more formal specification for coin selection, but that shouldn't necessarily restrict the way we write this document.